server/src/main/java/org/opensearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -316,60 +317,85 @@ && isTargetSameHistory()
+                    prepareTargetForTranslog(countNumberOfHistoryOperations(startingSeqNo), prepareEngineStep);
+                }, onFailure);
+
+                prepareEngineStep.whenComplete(prepareEngineTime -> {
+                    assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[phase2]");
+                    /*
+                     * add shard to replication group (shard will receive replication requests from this point on) now that engine is open.
+                     * This means that any document indexed into the primary after this will be replicated to this replica as well
+                     * make sure to do this before sampling the max sequence number in the next step, to ensure that we send
+                     * all documents up to maxSeqNo in phase2.
+                     */
+                    RunUnderPrimaryPermit.run(
+                        () -> shard.initiateTracking(request.targetAllocationId()),
+                        shardId + " initiating tracking of " + request.targetAllocationId(),
+                        shard,
+                        cancellableThreads,
+                        logger
+                    );
 
-                // we have to capture the max_seen_auto_id_timestamp and the max_seq_no_of_updates to make sure that these values
-                // are at least as high as the corresponding values on the primary when any of these operations were executed on it.
-                final long maxSeenAutoIdTimestamp = shard.getMaxSeenAutoIdTimestamp();
-                final long maxSeqNoOfUpdatesOrDeletes = shard.getMaxSeqNoOfUpdatesOrDeletes();
-                final RetentionLeases retentionLeases = shard.getRetentionLeases();
-                final long mappingVersionOnPrimary = shard.indexSettings().getIndexMetadata().getMappingVersion();
-                phase2(
-                    startingSeqNo,
-                    endingSeqNo,
-                    phase2Snapshot,
-                    maxSeenAutoIdTimestamp,
-                    maxSeqNoOfUpdatesOrDeletes,
-                    retentionLeases,
-                    mappingVersionOnPrimary,
-                    sendSnapshotStep
-                );
+                    final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("snapshot translog for recovery; current size is [{}]", countNumberOfHistoryOperations(startingSeqNo));
+                    }
+                    final Translog.Snapshot phase2Snapshot = shard.newChangesSnapshot(
+                        PEER_RECOVERY_NAME,
+                        startingSeqNo,
+                        Long.MAX_VALUE,
+                        false,
+                        true
+                    );
+                    resources.add(phase2Snapshot);
+                    retentionLock.close();
```

Comment:
From what i understand this is what we are skipping for remote translog. Is that correct?

Reply:
Yes

Reply:
The snapshot is coming not from translog operations but Lucene operations. Do we still need this?

Reply:
>The snapshot is coming not from translog operations but Lucene operations. Do we still need this?

This is skipped when remote translog is there. [code](https://github.com/opensearch-project/OpenSearch/pull/4301/files#diff-5ddc86ca495e5c853fe486c7f5d4ad9a3b2562356e912a287b4b37e889edd3eaR323)

> In case you are working on removing all operation based recovery, then do we need to handle PRRL altogether differently?

That is correct. We dont need PRRL for replica recovery, but we would still need this for primary-primary recovery.

Reply:
Yes